### ISSUE TITLE
[Feature Request] Allow to restore tables in a live cluster

### ISSUE THREAD

**Issue by cscetbon** [instaclustr/cassandra-backup#17]
In order for [CassKop](https://github.com/Orange-OpenSource/casskop) to start using the sidecar we would need to be able to restore tables in a live cluster. The restore should be done in the following steps:
- restore the data in a temporary folder
- truncate all the tables that gonna be restored
- create hardlinks to the restored files in the target folders
- do a nodetool refresh (jmx calls)
- clean temporary folder

**Comment by smiklosovic** [comment]
@cscetbon 

The approach seems all fine in case there are no changes of a schema. What if this is the case:

* in time t1 with schema s1 I do a backup
* in time t2, I alter a table, e.g. I add a column c1, so this will be on schema s2
* in time t3, I want to make the restore, if I apply your procedure, is not it the case that I would restore SSTables on schema s1 into (empty, as truncated) table with schema s2?

**Comment by cscetbon** [comment]
The procedure I described would only work if the schema is identical and the topology didn't change.
If topology changed, you'll need to restore it using [sstableloader](https://docs.datastax.com/en/dse/5.1/dse-admin/datastax_enterprise/tools/toolsSStables/toolsBulkloader.html) cause the distribution/replication can be totally different.
If the schema changed, there is no way you can restore the table unless you restore it as it was using the previous method and apply the schema change manually, but then what about the data contained in that new column in your example ?
We won't be able to resolve all cases but we can try to get as close as possible 😉
@ahmedjami don't hesitate to weigh in on the procedure I described if I forgot to mention anything.

**Comment by smiklosovic** [comment]
@cscetbon yes this is what I was thinking, however, I am not sure this tool should be enriched to deal with sstableloader, I feel like we are mixing things here. Not saying your idea with sstableloader is not sound, it is just I dont see its place in this particular piece of tooling, maybe we could do something custom or similar or make it another module. 

As a food for thought, sstableloader might be embedded into our CLI tool, if you check the code it is just calling some stuff underneath so we could embed it and call however, if we are restoring e.g. 5 nodes cluster, this means we would have to download all data to one place and stream it by loader back to cluster? Or data would be saved locally to that node and then we would use sstableloader just for that part and this would be repeated for each node?

**Comment by cscetbon** [comment]
I only talked about sstableloader because you asked how it should work in cases where the live restore would not work. It’s a case we should handle though but I don’t know what form it should take. I also had in mind to embed it in your tool but this is something we should talk about. 
To answer your question, you don’t have to push everything to one place to load sstables, you can always ask different nodes/containers to load different sstables using sstableloader.
But I don’t think this part should be the goal of another ticket. 
The current ticket is to restore tables on a live cluster where schema and topology are identical

**Comment by ahmedjami** [comment]
Hi,
@cscetbon the procedure is good to restore tables without a restart.
Going through a temporary folder will be very useful with the new nodetool command " import <ks> <cf> **-d <directory>**" when we will upgrade to C*4.  It will replace the actual nodetool refresh.

@smiklosovic yes "refresh" performs loading of new sstables without checking schema. So, between t1 and t3 in your example, your data will be well restored. I just tested it to be sure and it works in both cases: drop or add  column.
Refresh just needs that replication and token ownership are the same for every node. Otherwise, an sstabeloder will be required.

**Comment by smiklosovic** [comment]
@ahmedjami yes, I can already incorporate this into backup tooling so it would autodetect that it runs against Cassandra 4 so `nodetool import` would be invoked instead of making hardlinks and refreshing so from the user's point of view it will be all transparent. 

However, to `truncate` means what we need to execute a CQL command and this tool does not know it yet but that is just an implementation matter. You would have to be sure that you propagate all your ssl stuff to your containers or what have you, sidecar (where backup-restore is embedded into) knows how to extract these credentials dynamically from Kubernetes secrets, consult this documentation to know more (1), I will do this the same way.

(1) https://github.com/instaclustr/cassandra-operator/wiki/Accounts,-login-and-SSL#ssl-with-sidecar

**Comment by cscetbon** [comment]
@smiklosovic not really, we can use jmx https://github.com/apache/cassandra/blob/cassandra-3.11/src/java/org/apache/cassandra/service/StorageServiceMBean.java#L447-L457

**Comment by smiklosovic** [comment]
@cscetbon nice! I was looking into this interface looking for this method but I had to be blind. Thank you, this simplifies things a lot.

**Comment by smiklosovic** [comment]
@cscetbon @ahmedjami I am struggiling to understand the workflow you have described at the 
beginning. Lets revisit that:

* restore the data in a temporary folder
* truncate all the tables that gonna be restored
* create hardlinks to the restored files in the target folders
* do a nodetool refresh (jmx calls)
* clean temporary folder

This is all nice if we are ever going to deal with _one_ node only. Imagine your cluster is consisting of 5 nodes. To truncate a table means that we invoke respective JMX method. That is fine, but you can not _truncate per node_. If you have 5 nodes in a ring and you truncate on the first node, it means that data would be truncated on all other nodes. Hence if you did all these steps and you would go to restore the second node, truncating it again would truncate what you restored on the first node.

Docs on truncate say:

```
    /**
     * Truncates (deletes) the given table from the provided keyspace.
     * Calling truncate results in actual deletion of ALL DATA IN THE CLUSTER
     * under the given table and IT WILL FAIL UNLESS ALL HOSTS ARE UP.
     * All data in the given column family will be deleted, but its definition
     * will not be affected.
```

So, I would change the flow like this:

* restore the data in a temporary folder **on each node**
* truncate all the tables that gonna be restored **calling truncate on the first node**, this means that data to restore will be downloaded and there will be no data in each node after calling truncate
* create hardlinks to the restored files in the target folders **on each node**
* do a nodetool refresh (jmx calls) **on each node**
* clean temporary folder **on each node**

If these steps would be for one node in a cluster only, that is all fine and I can execute it in one flow, however I am not sure how to do this if we have not only one node in a cluster.

There might be a flag / option if a particular restoration request sent to node / invoked by CLI should truncate a table or not and you would have to orchestrate this from your operator.

**Comment by cscetbon** [comment]
yes @smiklosovic, sorry that was implicit for the truncate. I had exactly the same idea. You can choose the first node or a random node by providing a flag. the only thing is the truncate must happen before you restore on each node. So you need to have a way to ensure it's done sequentially.

**Comment by smiklosovic** [comment]
@cscetbon 

> So you need to have a way to ensure it's done sequentially.

That is your task here, to orchestrate this from the operator. This tool is going to be invoked per  node only, if you are going to call via Sidecar, a REST request would have to be sent. If you are invoking it via CLI, you would have to manage that again, per node.

From the implementation perspective, both approaches should same, but this orchestration of what to do when have to be done by you. This tool is "cluster agnostic", it cares about one node at a time at most.

**Comment by smiklosovic** [comment]
@cscetbon one more idea,

it could be done _by me_ but for you to ingest this whole project more easilly, I would make one "cluster restoration" endpoint in our sidecar and this endpoint would do this whole orchestration on your behalf so your only entry point to restoration would be a sidecar of any node which would act like a "coordinator".

The input of this sidecar endpoint would be (among other things) a mapping between remote backup and local node (as data are stored again per node). Basically it would be a list of restoration requests and sidecar would take care of this internally - it would internally call all other sidecars with specific restoration requests to that respective node. 

How do you actually want to invoke this restoration?

**Comment by ahmedjami** [comment]
Basicaly: when a truncate has been invoked by a first node/sidecar, no data files will are present on the rest of the nodes so you can check this condition before invoke a truncate.
But, this could make sense only with sequential node restore to avoid 2 truncate on the same time.

Otherwise, in order to keep the sidecar generic and responsible off all the procedure without giving the control of truncate to an external programme/operator. I vote for the second idea @smiklosovic

**Comment by smiklosovic** [comment]
@ahmedjami good, I am about to finish _importing_ strategy as discussed for Cassandra 4, once it is done, the "old" way of doing this (by hardlinks) should logically more or less same ... stay tuned!

**Comment by cscetbon** [comment]
Good to hear that @smiklosovic. you'll have to tell us a bit more about this coordination thing I'm not sure to fully understand. But I like having only one endpoint for the whole thing if you can make it work ;) I'm not sure I understand why you need this mapping. Can you elaborate ?

**Comment by cscetbon** [comment]
If it is a sidecar that acts as a coordinator you need to be sure to get everything you need to contact the right sidecar. I feel like that's what you said but I'll need to know more about it. Cause CassKop has a notion of datacenters that you don't have for instance. I just wanna be sure you contact the right sidecars. Maybe you'll tell me we just have to treat different [datacenters](https://github.com/Orange-OpenSource/casskop/blob/2944ff30e3cf42047d5b5087478bba625611d58c/documentation/description.md#cassandra-notion-of-dc-and-racks) as different groups of sidecars with one coordinator in each ?

**Comment by smiklosovic** [comment]
@cscetbon interesting point about the datacenters! I will do it with that in mind. I ll try to do some proof of concept and I ll do some docs about.

What I meant, in a nutshell, is that if there is a sidecar chosen to be a coordinator a "datacenter restore request" is sent to, that very sidecar has to know what other sidercars (except itself) are in that particular datacenter.

We can be smart about this and calling other JMX method to see what other nodes are there so all other sidecars would be on same IP address but I do not know yet how _stable_ this would be. Or we can just play it simple and a list of sidecars would be given in a restore request by operator so this burden of realizing your topology would be left on you.

I ll try to investigate the first option and restore would work just on a datacenter level, not cluser wide level, you would have to issue another restoration request for the other dc by choosing some other sidecar from there to restore that dc ... lets just keep it simple here.
